File: 2033. Minimum Operations to Make a Uni-Value Grid/main.py
from typing import List


class Solution:
    def minOperations(self, grid: List[List[int]], x: int) -> int:
        # 1. Check all remainders
        # 2. Flatten and sort the input
        # 3. Prefix sum / suffix sum

        for row in grid:
            for n in row:
                if n % x != grid[0][0] % x:
                    return -1
        
        nums = sorted([n for row in grid for n in row])

        prefix = 0
        total = sum(nums)
        res = float("inf")

        for i in range(len(nums)):
            cost_left = nums[i] * i - prefix
            cost_right = total - prefix - nums[i] * (len(nums) - i)
            res = min(res, (cost_left + cost_right) // x)
            prefix += nums[i]
        
        return res
        

File: 2033. Minimum Operations to Make a Uni-Value Grid/main.go
package main

import (
	"math"
	"slices"
)

func minOperations(grid [][]int, x int) int {
	nums := make([]int, 0)
	total := 0

	for _, row := range grid {
		for _, n := range row {
			if n%x != grid[0][0]%x {
				return -1
			}
			nums = append(nums, n)
			total += n
		}
	}

	slices.Sort(nums)

	prefix := 0
	res := math.MaxInt

	for i := range nums {
		costLeft := nums[i]*i - prefix
		costRight := total - prefix - (nums[i] * (len(nums) - i))
		operations := (costLeft + costRight) / x
		res = min(res, operations)
		prefix += nums[i]
	}

	return res
}
